### ISSUE TITLE
Invalidating the cache

### ISSUE THREAD

**Issue by mxstbr** [apollographql/apollo-cache-control-js#1]
How do I invalidate the cache if e.g. a mutation comes in and changes something?

**Comment by martijnwalraven** [comment]
There is currently no way to invalidate the cache (other than waiting for cache entries to expire). We have some ideas that we'd like to implement as part of this spec, but it will take time to figure out the best approach, and we'd like to work with the community on that. I hope to have something to share as a starting point for discussion soon.

(Sorry for the late response, but I got bitten by the fact that creating a repo doesn't automatically make you watch it.)

**Comment by mxstbr** [comment]
Hey, any news here? :blush: We really need this.

**Comment by martijnwalraven** [comment]
No news, but it would be helpful to hear more about your use cases. What granularity of invalidation would you need? Would invalidations be a result of mutations, or would you also require out of band invalidation?

**Comment by mxstbr** [comment]
Just mutations right now, basically we need to be able to say "Hey a new message was sent in this thread so invalidate the cache for the thread" and stuff like that.

**Comment by mxstbr** [comment]
Is there any news here @martijnwalraven? We'd love love love to start using caching, but our data is all very real-time so without invalidation we can't :cry:

**Comment by martijnwalraven** [comment]
It's on the roadmap, and we've had some fruitful design sessions, but an implementation in Engine will probably be at least several months away.

**Comment by mxstbr** [comment]
Oh well, at least it's going to happen! 👍

**Comment by wux5** [comment]
Hi @mxstbr , so before the cache invalidation actually comes out, how do you handle the issue right now? I'm in a similar situation that I simply want to invalidate a cached query in some cases, so that it'll be refetched whenever it runs next time. @martijnwalraven any advice how to achieve this? Thanks.

**Comment by mxstbr** [comment]
We unfortunately can't cache until we can invalidate the cache, so I don't handle it at all :confused:

**Comment by phpthug** [comment]
Any updates on the timing of introducing cache invalidation? I am looking at needing it for mutations right now, same as @mxstbr.

**Comment by M-Zuber** [comment]
I'll be a third vote for this,
For our use case, any sort of access to the data in the cache would be useful to just manually apply the mutation onto cache in parallel with the actual request

**Comment by Tehnix** [comment]
Are the cache keys predictable? If so, one could imagine using Memcache and then be able to invalidate the cache from whatever service needs to do this, via directly invalidating it in the Memcache instance that Apollo Engine uses.

Alternatively, submitting a mutation that invalidates it, that the service itself ignores?

I'd love to hear what people have done here, because there's simply no way we can use a cache that does not support invalidation :/

**Comment by danj3** [comment]
I think this works, add a parameter to the mutation that is a counter managed by the client and the server ignores. Increment it each time you want to invalidate the cache, since the parameter will change.

**Comment by Tehnix** [comment]
It's not so much cache invalidation from the client, but from the server side of things. Imagine that you choose to cache the description of an image. Now, this image gets changed from some other part of the system that is not via a Graphql endpoint, and doesn't perform a mutation. How then do we let the apollo cache know that it needs to invalidate this cache.

Sorta what memcache/redis is normally used for, where you'd just request from cache, and whoever updates the data, is responsible for clearing out that item from the cache.

**Comment by evans** [comment]
I'm going to close this issue for now, since we're moving this repository into apollographql/apollo-server#1294.

In terms of an update, we're working on an invalidating cache apollographql/apollo-server/pull/1192 that will be used for the full response caching and datasource caching inside of apollo-server. Commenting on apollographql/apollo-server/1050 or opening another issue specifically about invalidation would be extremely helpful.

A question on my mind is how do you want the cache invalidation to happen? Are you looking to invalidate operations by name or something more granular?

Please feel free to comment on the PR or in an issue
